File: Day_7/main.py
class directory:
    _name = ""
    _root = None
    _parent = None
    _subdirs = []
    _files = []

    def __init__(self, name, parent = None, root = None):
        self._name = name
        self._parent = parent
        self._subdirs = []
        self._files = []
        if root == None:
            self._root = self
        else:
            self._root = root
    
    def ls(self, lines):
        for line in lines:
            args = line.split(" ")
            if (args[0] == "dir"):
                exists = False
                for sd in self._subdirs:
                    if sd._name == args[1]:
                        exists = True
                        print("Directory " + str(args[1]) + " already exists")
                        break
                if not exists:
                    self._subdirs.append(directory(args[1], self, self._root))
            else:
                exists = False
                for file in self._files:
                    if file[0] == args[1]:
                        exists = True
                        print("File " + str(args[1]) + " already exists")
                        break
                if not exists:
                    self._files.append([args[1], int(args[0])])
    
    def cd(self, dirname):
        if dirname == "/":
            return self._root
        if dirname == "..":
            if self._parent != None:
                return self._parent
            else:
                print("No parent directory")
        else:
            for subdir in self._subdirs:
                if subdir._name == dirname:
                    return subdir
            print(dirname + ": No such directory")
        return self
    
    def get_size(self):
        size = 0
        for file in self._files:
            size += int(file[1])
        for subdir in self._subdirs:
            size += subdir.get_size()
        return size
    
    def get_part1(self):
        res = self.get_size()
        if (res > 100000):
            res = 0
        for subdir in self._subdirs:
            res += subdir.get_part1()
        return res
    
    def get_part2(self, to_free):
        res = 0
        size = self.get_size()
        if (size < to_free):
            res = 0
        else:
            res = size
        for subdir in self._subdirs:
            subdir_res = subdir.get_part2(to_free)
            if (subdir_res < res and subdir_res > 0):
                res = subdir_res
        return res

def main():
    ls = []
    root = directory("/")
    curent = root
    with open("input.txt") as f:
        lines = f.read().splitlines() 
        for line in lines:
            if line == "\n":
                break
            args = line.split(" ")
            if (args[0] == "$"):
                if len(ls) > 0:
                    curent.ls(ls)
                    ls = []
                if (args[1] == "cd"):
                    curent = curent.cd(args[2])
            else:
                ls.append(line)
        if (len(ls) > 0):
            curent.ls(ls)
            ls = []
    print("Part 1 res: " + str(root.get_part1()))
    print("Part 2 res: " + str(root.get_part2(root.get_size() - 40000000)))  
    return (0)

if __name__ == "__main__":
    main()

File: Day_7/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <utility>

class directory
{
    public :
        directory( std::string name, directory * parent = NULL ) : _name(name), _parent(parent) {};
        ~directory() {
            _subdirs.clear();
            _files.clear();
        };

        void ls( std::vector< std::string > lines )
        {
            for ( std::vector< std::string >::iterator it = lines.begin(); it < lines.end(); it++ )
            {
                std::string line = *it;
                std::string type = line.substr(0, line.find_first_of(" "));
                if ( type == "dir" )
                {
                    std::string name = line.substr(line.find_first_of(" ") + 1);
                    bool exists = false;
                    for ( std::vector< directory >::iterator it = _subdirs.begin(); it < _subdirs.end(); it++ )
                    {
                        if ( it->get_name() == name )
                        {
                            exists = true;
                            break;
                        }
                    }
                    if ( !exists )
                        _subdirs.push_back( directory( name, this ) );
                }
                else
                {
                    std::string name = line.substr(line.find_first_of(" ") + 1, line.find_last_of(" "));
                    int size = std::stoi( line.substr( 0, line.find_first_of(" ") ) );
                    bool exists = false;
                    for ( std::vector< std::pair< std::string, int > >::iterator it = _files.begin(); it < _files.end(); it++ )
                    {
                        if ( it->first == name )
                        {
                            exists = true;
                            break;
                        }
                    }
                    if ( !exists )
                        _files.push_back( std::make_pair( name, size ) );
                }
            }
        }

        directory * cd( std::string name )
        {
            if ( name == ".." )
            {
                return _parent;
            }
            else
            {
                for ( std::vector< directory >::iterator it = _subdirs.begin(); it < _subdirs.end(); it++ )
                {
                    if ( it->get_name() == name )
                    {
                        return &(*it);
                    }
                }
                std::cout << "Directory " << name << " does not exist" << std::endl;
            }
            return NULL;
        }

        std::string get_name() { return _name; };
        directory * get_parent() { return _parent; };
        std::vector< directory > get_subdirs() { return _subdirs; };
        std::vector< std::pair< std::string, int > > get_files() { return _files; };

        int get_size()
        {
            int size = 0;
            for ( std::vector< std::pair< std::string, int > >::iterator it = _files.begin(); it < _files.end(); it++ )
            {
                size += it->second;
            }
            for ( std::vector< directory >::iterator it = _subdirs.begin(); it < _subdirs.end(); it++ )
            {
                size += it->get_size();
            }
            return size;
        }
    
        void print_tree( std::string prefix = "- " )
        {
            std::cout << prefix << _name << " (dir)" << std::endl;
            prefix = "  " + prefix;
            for ( std::vector< std::pair< std::string, int > >::iterator it = _files.begin(); it < _files.end(); it++ )
            {
                std::cout << prefix << it->first << " (file, size=" << it->second << std::endl;
            }
            for ( std::vector< directory >::iterator it = _subdirs.begin(); it < _subdirs.end(); it++ )
            {
                it->print_tree( prefix );
            }
        }

        int get_part1()
        {
            int res = 0;
            int size = this->get_size();
            if ( size <= 100000 )
                res += size;
            for ( std::vector< directory >::iterator it = _subdirs.begin(); it < _subdirs.end(); it++ )
                res += it->get_part1();
            return res;
        }

        int get_sum()
        {
            int res = 0;
            for ( std::vector< std::pair< std::string, int > >::iterator it = _files.begin(); it < _files.end(); it++ )
            {
                res += it->second;
            }
            for ( std::vector< directory >::iterator it = _subdirs.begin(); it < _subdirs.end(); it++ )
            {
                res += it->get_sum();
            }
            return res;
        }

        int get_part2( int to_free )
        {
            int dir_size = this->get_size();
            int res = 0;
            if ( dir_size >= to_free )
                res = dir_size;
            for ( std::vector< directory >::iterator it = _subdirs.begin(); it < _subdirs.end(); it++ )
            {
                int subdir_res = it->get_part2( to_free );
                if ( subdir_res >= to_free && subdir_res > 0)
                {
                    res = subdir_res;
                }
            }
            return res;
        }
    
    private :
        std::string _name;
        directory * _parent;
        std::vector< directory > _subdirs;
        std::vector< std::pair< std::string, int > > _files;
};

int main()
{
    directory root( "/" );
    directory * current = &root;
    std::vector< std::string > args;
    std::string line;
    std::ifstream myfile1 ("input.txt");
    if (myfile1.is_open())
    {
        while ( getline (myfile1,line) )
        {
            if ( line[0] == '$')
            {
                if ( args.size() )
                {
                    current->ls( args );
                    args.clear();
                }
                line = line.substr( 2 );
                std::string cmd = line.substr(0, line.find_first_of(" "));
                if ( cmd == "cd" )
                {
                    directory * next;
                    if ( line.find_first_of(" ") == std::string::npos || line.substr( line.find_first_of(" ") + 1 ) == "/" )
                    {
                        current = &root;
                        continue;
                    }
                    else
                        next = current->cd( line.substr( line.find_first_of(" ") + 1 ) );
                    if ( next )
                        current = next;
                }
            }
            else
            {
                args.push_back( line );
            }
        }
        if ( args.size() )
        {
            current->ls( args );
            args.clear();
        } 
        std::cout << "Part 1 res: " << root.get_part1() << std::endl;
        int to_free = root.get_sum() - 40000000;
        std::cout << "Part 2 res: " << root.get_part2( to_free ) << std::endl;

    } 
    myfile1.close();

    return (0);
}